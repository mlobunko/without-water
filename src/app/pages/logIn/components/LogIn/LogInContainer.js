import React from 'react';
import { connect } from 'react-redux';
import { Redirect, withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';

import * as model from '../../model';
import * as actions from '../../actions';
import * as constants from '../../constants';
import LogInComponent from './LogInComponent';
import { REPO_NAME } from 'main/constants';

import authorized from 'shared/authorized';
import fetching from 'shared/fetching';
import error from 'shared/error';

export class LogInContainer extends React.Component {
  state = {
    email: '',
    password: '',
    emailError: '',
    passwordError: ''
  };

  componentDidMount = () => {
    this.props.errorActionsSetNull();
  };

  static getDerivedStateFromProps(props) {
    if (!!props.errorState.loginError) {
      return {
        password: ''
      };
    } else return null;
  }

  setStateErrorsToNull = () => {
    this.setState({
      emailError: '',
      passwordError: ''
    });
  };

  handleInputsChange = e => {
    if (this.props.errorState.isError) {
      this.props.errorActionsSetNull();
    }
    if (!!this.state.emailError || !!this.state.passwordError) {
      this.setStateErrorsToNull();
    }
    const value = e.currentTarget.value;
    const idTarget = e.currentTarget.id;
    this.setState({
      [idTarget]: value
    });
  };
  handleSubmit = e => {
    e.preventDefault();
    const validator = model.validator(this.state.email, this.state.password);
    if (validator.email && validator.password) {
      this.props.actionsSubmit({
        email: '' + this.state.email,
        password: '' + this.state.password
      });
    } else {
      if (!validator.email) {
        this.setState({
          emailError: constants.ERROR_FORM_LIB.email
        });
      }
      if (!validator.password) {
        this.setState({
          passwordError: constants.ERROR_FORM_LIB.password
        });
      }
    }
  };
  onKeyPress(event) {
    if (event.which === 13 /* Enter */) {
      event.preventDefault();
    }
  }

  render() {
    const { from } = this.props.location.state || {
      from: { pathname: `${REPO_NAME}/` }
    };
    if (this.props.authorized) {
      return <Redirect to={from} />;
    }

    return (
      <LogInComponent
        email={this.state.email}
        emailError={this.state.emailError}
        fetching={this.props.fetching}
        handleInputsChange={this.handleInputsChange}
        handleSubmit={this.handleSubmit}
        isError={this.props.errorState.isError}
        loginError={this.props.errorState.loginError}
        password={this.state.password}
        passwordError={this.state.passwordError}
        onKeyPress={this.onKeyPress}
        serverError={this.props.errorState.serverError}
      />
    );
  }
}

const mapStateToProps = state => ({
  authorized: authorized.selectors.getAll(state),
  fetching: fetching.selectors.getAll(state),
  errorState: error.selectors.getAll(state)
});

const mapDispatchToProps = {
  actionsSubmit: actions.submit,
  errorActionsSetNull: error.actions.setNull
};

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(LogInContainer)
);

LogInContainer.propTypes = {
  authorized: PropTypes.bool.isRequired,
  fetching: PropTypes.bool.isRequired,
  errorState: PropTypes.shape({
    isError: PropTypes.bool.isRequired,
    loginError: PropTypes.string,
    serverError: PropTypes.string
  })
};
