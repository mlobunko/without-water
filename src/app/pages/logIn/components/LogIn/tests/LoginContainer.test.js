import React from 'react';
import { shallow } from 'enzyme';

import { LogInContainer } from '../LogInContainer';

const initialProps = {
  authorized: false,
  fetching: false,
  errorState: {
    isError: false,
    loginError: '',
    serverError: ''
  },
  location: {
    state: '/'
  },
  actionsSubmit: jest.fn(),
  errorActionsSetNull: jest.fn()
};

const initialState = {
  email: '',
  password: '',
  emailError: '',
  passwordError: ''
};
const initialE = {
  currentTarget: {
    value: 'fa@fa.com',
    id: 'email'
  },
  preventDefault: jest.fn(),
  which: 13
};

describe('componentDidMount', () => {
  const props = {
    ...initialProps
  };
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('set errors to null', () => {
    shallow(<LogInContainer {...props} />);
    expect(props.errorActionsSetNull).toHaveBeenCalledTimes(1);
  });
});

describe('getDerivedStateFromProps', () => {
  const props = {
    ...initialProps
  };

  it('should set password value to empty string', () => {
    const wrapper = shallow(<LogInContainer {...props} />);
    wrapper.setState({
      ...initialState,
      password: '123456'
    });
    wrapper.setProps({
      ...props,
      errorState: {
        isError: true,
        loginError: 'error',
        serverError: ''
      }
    });
    expect(wrapper.state().password).toEqual('');
  });

  it('should not set password value to empty string', () => {
    const wrapper = shallow(<LogInContainer {...props} />);
    wrapper.setState({
      ...initialState,
      password: '123456'
    });
    wrapper.setProps({
      ...props,
      errorState: {
        isError: true,
        loginError: '',
        serverError: 'error'
      }
    });
    expect(wrapper.state().password).toEqual('123456');
  });
});

describe('setStateErrosToNull', () => {
  const props = {
    ...initialProps
  };
  it('return state with empty errors strings', () => {
    const wrapper = shallow(<LogInContainer {...props} />);
    const newState = {
      ...initialState,
      emailError: 'error',
      passwordError: 'error'
    };
    const expectedState = {
      ...initialState
    };
    wrapper.setState(newState);
    wrapper.instance().setStateErrorsToNull();
    expect(wrapper.state()).toEqual(expectedState);
  });
});

describe('handleInputsChange', () => {
  const e = {
    currentTarget: {
      value: 'fa@fa.com',
      id: 'email'
    }
  };
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('set errors login and server to empty strings', () => {
    const props = {
      ...initialProps
    };
    const newProps = {
      ...initialProps,
      errorState: {
        isError: true,
        loginError: 'error',
        serverError: 'error'
      }
    };
    const wrapper = shallow(<LogInContainer {...props} />);
    jest.resetAllMocks();
    wrapper.setProps(newProps);
    wrapper.instance().handleInputsChange(e);
    expect(props.errorActionsSetNull).toHaveBeenCalledTimes(1);
  });

  it('set state errors to empty stings', () => {
    const props = {
      ...initialProps
    };
    const state = {
      ...initialState,
      emailError: 'error',
      passwordError: 'error'
    };
    const wrapper = shallow(<LogInContainer {...props} />);
    wrapper.instance().setStateErrorsToNull = jest.fn();
    wrapper.setState(state);
    wrapper.instance().handleInputsChange(e);
    expect(wrapper.instance().setStateErrorsToNull).toHaveBeenCalledTimes(1);
  });

  it('set new value without the call error to empty strings', () => {
    const props = {
      ...initialProps
    };
    const expectedState = {
      ...initialState,
      email: 'fa@fa.com'
    };
    const wrapper = shallow(<LogInContainer {...props} />);
    wrapper.instance().handleInputsChange(e);
    expect(wrapper.state()).toEqual(expectedState);
  });
});

describe('handleSubmit', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('call preventDefault', () => {
    const props = {
      ...initialProps
    };
    const e = {
      ...initialE
    };
    const wrapper = shallow(<LogInContainer {...props} />);
    wrapper.instance().handleSubmit(e);
    expect(e.preventDefault).toHaveBeenCalledTimes(1);
  });

  describe('call actionSubmut', () => {
    const e = {
      ...initialE
    };
    const props = {
      ...initialProps
    };
    const expectedParameters = {
      email: 'fa@fa.com',
      password: '123456'
    };
    const state = {
      ...initialState,
      ...expectedParameters
    };

    it('call actionsSubmit', () => {
      const wrapper = shallow(<LogInContainer {...props} />);
      wrapper.setState(state);
      wrapper.instance().handleSubmit(e);
      expect(props.actionsSubmit).toHaveBeenCalledWith(expectedParameters);
    });

    it('not set email error', () => {
      const wrapper = shallow(<LogInContainer {...props} />);
      wrapper.setState(state);
      wrapper.instance().handleSubmit(e);
      expect(wrapper.state().emailError).toEqual('');
    });

    it('not set password error', () => {
      const wrapper = shallow(<LogInContainer {...props} />);
      wrapper.setState(state);
      wrapper.instance().handleSubmit(e);
      expect(wrapper.state().passwordError).toEqual('');
    });
  });
  describe('set email error', () => {
    beforeEach(() => {
      jest.resetAllMocks();
    });
    const e = {
      ...initialE
    };
    const props = {
      ...initialProps
    };
    const state = {
      ...initialState,
      email: '213',
      password: '123456'
    };

    it('set email error', () => {
      const wrapper = shallow(<LogInContainer {...props} />);
      wrapper.setState(state);
      wrapper.instance().handleSubmit(e);
      expect(wrapper.state().emailError).not.toEqual('');
    });

    it('not call actionsSubmit', () => {
      const wrapper = shallow(<LogInContainer {...props} />);
      wrapper.setState(state);
      wrapper.instance().handleSubmit(e);
      expect(props.actionsSubmit).toHaveBeenCalledTimes(0);
    });
  });

  describe('set password error', () => {
    beforeEach(() => {
      jest.resetAllMocks();
    });
    const e = {
      ...initialE
    };
    const props = {
      ...initialProps
    };
    const state = {
      ...initialState,
      email: 'fa@fa.com',
      password: '12'
    };

    it('set password error', () => {
      const wrapper = shallow(<LogInContainer {...props} />);
      wrapper.setState(state);
      wrapper.instance().handleSubmit(e);
      expect(wrapper.state().passwordError).not.toEqual('');
    });

    it('not call actionsSubmit', () => {
      const wrapper = shallow(<LogInContainer {...props} />);
      wrapper.setState(state);
      wrapper.instance().handleSubmit(e);
      expect(props.actionsSubmit).toHaveBeenCalledTimes(0);
    });
  });
});

describe('onKeyPress', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });
  it('call preventDefault', () => {
    const e = {
      ...initialE
    };
    const props = {
      ...initialProps
    };
    const wrapper = shallow(<LogInContainer {...props} />);
    wrapper.instance().onKeyPress(e);
    expect(e.preventDefault).toHaveBeenCalledTimes(1);
  });
});

describe('render', () => {
  const props = {
    ...initialProps
  };
  it('renders correctly', () => {
    const wrapper = shallow(<LogInContainer {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
