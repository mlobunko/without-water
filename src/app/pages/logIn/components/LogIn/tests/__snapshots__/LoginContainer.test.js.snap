// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render renders correctly 1`] = `
<LogInComponent
  email=""
  emailError=""
  fetching={false}
  handleInputsChange={[Function]}
  handleSubmit={[Function]}
  isError={false}
  loginError=""
  onKeyPress={[Function]}
  password=""
  passwordError=""
  serverError=""
/>
`;
